Don't repeat suggestions for same symbol

ire4ever1190 opened the pull request:
Tracks seen symbols while iterating through all symbols to find similar spellings/suggestions. This is done to stop repeated suggestions for the same symbol occuring.

Bare bones example
```nim
echo l
```
`nim check --spellSuggest:4 test.nim`
```
test.nim(1, 6) Error: undeclared identifier: 'l'
candidates (edit distance, scope distance); see '--spellSuggest': 
 (1, 3): '$' [func declared in /home/jake/.choosenim/toolchains/nim-#devel/lib/system/dollars.nim(12, 8)]
 (1, 3): '$' [func declared in /home/jake/.choosenim/toolchains/nim-#devel/lib/system/dollars.nim(31, 8)]
 (1, 3): '$' [func declared in /home/jake/.choosenim/toolchains/nim-#devel/lib/system/dollars.nim(31, 8)]
 (1, 3): '$' [func declared in /home/jake/.choosenim/toolchains/nim-#devel/lib/system/dollars.nim(31, 8)]
```
The bottom `$` is repeated three times

ire4ever1190 committed:
Track seen module graphs so symbols from the same module aren't repeated
Add test case

ire4ever1190 committed:
Track symbols instead of modules

metagn commented:
Shouldn't this be where `allSyms` is used in `suggest` instead of inside `allSyms` itself?

ire4ever1190 commented:
The version of `allSyms` (Different to one in `modulegraphs.nim`) that I modified is only used by `lookups.nim` and `suggest.nim`. By putting the changes inside `allSyms` I don't need to modify the multiple usages of it

ire4ever1190 commented:
Thought PR broke `nimsuggest` but was just that moving `nimsuggest` out of bin folder crashed it (Happens on devel also so think I'm just compiling wrong)

Araq commented on compiler/lookups.nim:
```diff
@@ -170,19 +170,30 @@ iterator allSyms*(c: PContext): (PSym, int, bool) =
+  # Track seen symbols so we don't duplicate them.
+  # The int is for the symbols name, and line info is
+  # to be able to tell apart symbols with same name but on different lines
+  var seen = initHashSet[(TLineInfo, int)]()
```
Why does TLineInfo matter here? We should only produce the first unique *name* as it's for typo corrections.

ire4ever1190 replied:
An example is if the user imports `options` and `httpclient`.
If they mistype `get` then it would be good to show them that there exists both a `get` for options and for httpclient.

ire4ever1190 replied:
We also need to return all symbols so `suggest.nim` can go through the symbols and match the correct one depending on the types the user has already written.

Araq replied:
But then this doesn't fix the issue that it suggests 10x `shl` etc

ire4ever1190 replied:
I could group them up into their modules to stop repeated suggestions?
e.g.
```nim
proc `shl`() = discard
proc shout() = discard

sh
```
Would give output
```
/tmp/test.nim(4, 1) Error: undeclared identifier: 'sh'
candidates; see '--spellSuggest':
 'shl':
    - /home/jake/.choosenim/toolchains/nim-#devel/lib/system/arithmetics.nim (first seen 158:6)
    - /tmp/test.nim (first seen 1:6)
 'shout': 
    - declared in test.nim (first seen 2:6)
```

This way `nimsuggest` still gets all the results and the console isn't spammed for every declaration of a symbol

Araq replied:
Just use different code paths for the typo checker and nimsuggest. I agree that nimsuggest must get all the results.

ire4ever1190 replied:
Now separate code paths. `nimsuggest` uses an iterator that builds on existing `allSyms` but only yields unique symbols (by line and name). The typo checker now skips symbols that are the same as the previous (They are sorted so its fairly efficient).

Remove line info from the typo suggestions. Since only the first declaration is shown it would be a bit misleading to the user if they couldn't see the other declarations ("Why is it only showing get in `options.nim` when I know its declared here also?" 

The output is a lot cleaner now. In an empty file if you type "sh" the output is like so
```
test.nim(1, 1) Error: undeclared identifier: 'sh'
candidates (edit distance, scope distance); see '--spellSuggest': 
 (1, 2): 'shl'
 (1, 2): 'shr'
```

ire4ever1190 committed:
Don't show duplicate symbols in spell checker

Removes the declared location from the message. Since we don't show duplicates anymore it would be a bit misleading if we only show the location for the first declaration of the symbol
